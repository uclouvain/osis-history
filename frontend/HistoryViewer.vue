<!--
  -
  -   OSIS stands for Open Student Information System. It's an application
  -   designed to manage the core business of higher education institutions,
  -   such as universities, faculties, institutes and professional schools.
  -   The core business involves the administration of students, teachers,
  -   courses, programs and so on.
  -
  -   Copyright (C) 2015-2021 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  -   This program is free software: you can redistribute it and/or modify
  -   it under the terms of the GNU General Public License as published by
  -   the Free Software Foundation, either version 3 of the License, or
  -   (at your option) any later version.
  -
  -   This program is distributed in the hope that it will be useful,
  -   but WITHOUT ANY WARRANTY; without even the implied warranty of
  -   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  -   GNU General Public License for more details.
  -
  -   A copy of this license - GNU General Public License - is available
  -   at the root of the source code of this program.  If not,
  -   see http://www.gnu.org/licenses/.
  -
  -->
<template>
  <div class="history-viewer">
    <div
        class="btn-group"
        data-toggle="buttons"
    >
      <label
          class="btn btn-default active"
          @click="mode = 'Table'"
      >
        <input
            type="radio"
            name="mode"
        >
        <span class="fa fa-table" />
      </label>
      <label
          class="btn btn-default"
          @click="mode = 'Timeline'; horizontal = false;"
      >
        <input
            type="radio"
            name="mode"
        >
        <span class="fas fa-ellipsis-v" />
      </label>
      <label
          class="btn btn-default"
          @click="mode = 'Timeline'; horizontal = true;"
      >
        <input
            type="radio"
            name="mode"
        >
        <span class="fa fa-ellipsis-h" />
      </label>
    </div>
    <div class="viewport">
      <span
          v-if="loading"
          class="spinner"
      />
      <span
          v-else-if="error"
          class="text-danger"
      >
        {{ error }}
      </span>
      <component
          :is="mode"
          v-else
          :entries="entries"
          :horizontal="horizontal"
      />
    </div>
  </div>
</template>

<script>
import Table from './components/Table';
import Timeline from './components/Timeline';

export default {
  name: 'HistoryViewer',
  components: {
    Table,
    Timeline,
  },
  props: {
    url: {
      type: String,
      required: true,
    },
    tags: {
      type: String,
      default: '',
    },
  },
  data () {
    return {
      mode: 'Table',
      horizontal: false,
      entries: [],
      error: '',
      loading: true,
    };
  },
  async mounted () {
    try {
      const response = await fetch(`${this.url}${this.tags ? '?tags=' + this.tags : ''}`);
      if (response.status === 200) {
        this.entries = await response.json();
      } else {
        this.error = response.statusText;
      }
    } catch (e) {
      this.error = e;
    }
    this.loading = false;
  },
};
</script>

<style>
.history-viewer .viewport {
  margin-top: 1em;
  overflow-x: scroll;
}

.history-viewer .spinner {
  display: block;
  margin: 0 auto;
  border: 3px solid #286090;
  border-top-color: transparent;
  border-radius: 50%;
  width: 20px;
  height: 20px;
  animation: spin 1.5s linear infinite;
}
</style>
