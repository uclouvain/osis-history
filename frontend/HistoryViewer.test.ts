/*
 *
 * OSIS stands for Open Student Information System. It's an application
 * designed to manage the core business of higher education institutions,
 * such as universities, faculties, institutes and professional schools.
 * The core business involves the administration of students, teachers,
 * courses, programs and so on.
 *
 * Copyright (C) 2015-2023 Universit√© catholique de Louvain (http://www.uclouvain.be)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of this license - GNU General Public License - is available
 * at the root of the source code of this program.  If not,
 * see http://www.gnu.org/licenses/.
 *
 */

import {beforeEach, expect, test} from 'vitest';
import {flushPromises, mount} from '@vue/test-utils';
import HistoryViewer from "./HistoryViewer.vue";
import type {Entry} from "./interfaces";
import fetchMock from "fetch-mock";

const apiUrl = "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/";
const mockEntries: Entry[] = [
  {
    author: 'John Doe',
    created: '06/04/2021 16:23',
    message: 'Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.',
  }, {
    author: 'Alice Smith',
    created: '03/04/2021 16:23',
    message: 'Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.\nSit facilis vitae sapiente. Placeat est qui et.',
  }, {
    author: 'Bob Martin',
    created: '01/04/2021 08:23',
    message: 'Corrupti <strong>reiciendis</strong> laboriosam repellat adipisci quis.\n\nIure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.',
  }, {
    author: 'John Doe',
    created: '28/03/2021 10:00',
    message: '',
  },
];

beforeEach(() => {
  fetchMock.restore().mock(apiUrl, mockEntries);
});

test('no history', async () => {
  fetchMock.restore().mock(apiUrl, []);
  expect(HistoryViewer).toBeTruthy();

  const wrapper = mount(HistoryViewer, {
    props: {url: apiUrl},
  });
  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});

test('http error', async () => {
  fetchMock.restore().mock(apiUrl, 404);

  const wrapper = mount(HistoryViewer, {
    props: {url: apiUrl},
  });
  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});


test('exception', async () => {
  fetchMock.restore().mock(apiUrl, {throws: new Error('Some network error')});

  const wrapper = mount(HistoryViewer, {
    props: {url: apiUrl},
  });
  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});

test('with history', async () => {
  const wrapper = mount(HistoryViewer, {
    props: {url: apiUrl},
  });

  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});

test('filtering on tag only', async () => {
  fetchMock.restore().mock(`${apiUrl}?tags=foo`, mockEntries);

  const wrapper = mount(HistoryViewer, {
    props: {
      url: apiUrl,
      tags: 'foo',
      defaultMode: 'vertical',
    },
  });

  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});

test('two modes with bad default', async () => {
  const wrapper = mount(HistoryViewer, {
    props: {
      url: apiUrl,
      modes: ['table', 'horizontal'],
      defaultMode: 'foo',
    },
  });

  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});

test('change modes', async () => {
  const wrapper = mount(HistoryViewer, {
    props: {url: apiUrl},
  });

  await flushPromises();
  const buttons = wrapper.findAll('.btn-default');
  expect(wrapper.findComponent({name: 'TableHistory'}).exists()).toBe(true);
  await buttons[1].trigger('click');
  expect(wrapper.findComponent({name: 'TimelineHistory'}).exists()).toBe(true);
  await buttons[0].trigger('click');
  expect(wrapper.findComponent({name: 'TableHistory'}).exists()).toBe(true);
  await buttons[2].trigger('click');
  expect(wrapper.findComponent({name: 'TimelineHistory'}).exists()).toBe(true);
});
