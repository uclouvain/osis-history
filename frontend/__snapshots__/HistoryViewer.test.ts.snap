// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exception 1`] = `
"<div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fas fa-ellipsis-v\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div>
<div class=\\"viewport\\"><span class=\\"text-danger\\">Some network error</span></div>"
`;

exports[`filtering on tag only 1`] = `
"<div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fas fa-ellipsis-v\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div>
<div class=\\"viewport\\">
  <div>
    <ul class=\\"timeline\\">
      <li class=\\"timeline-item\\">
        <div class=\\"timeline-badge\\"></div>
        <div class=\\"timeline-panel\\">
          <div class=\\"timeline-heading\\">
            <h3 class=\\"h4\\"><span class=\\"glyphicon glyphicon-time\\"></span> 06/04/2021 16:23 - <span class=\\"glyphicon glyphicon-user\\"></span> John Doe</h3>
          </div>
          <div class=\\"timeline-body\\">Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.</div>
        </div>
      </li>
      <li class=\\"timeline-item\\">
        <div class=\\"timeline-badge\\"></div>
        <div class=\\"timeline-panel\\">
          <div class=\\"timeline-heading\\">
            <h3 class=\\"h4\\"><span class=\\"glyphicon glyphicon-time\\"></span> 03/04/2021 16:23 - <span class=\\"glyphicon glyphicon-user\\"></span> Alice Smith</h3>
          </div>
          <div class=\\"timeline-body\\">Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.
            Sit facilis vitae sapiente. Placeat est qui et.</div>
        </div>
      </li>
      <li class=\\"timeline-item\\">
        <div class=\\"timeline-badge\\"></div>
        <div class=\\"timeline-panel\\">
          <div class=\\"timeline-heading\\">
            <h3 class=\\"h4\\"><span class=\\"glyphicon glyphicon-time\\"></span> 01/04/2021 08:23 - <span class=\\"glyphicon glyphicon-user\\"></span> Bob Martin</h3>
          </div>
          <div class=\\"timeline-body\\">Corrupti <strong>reiciendis</strong> laboriosam repellat adipisci quis.

            Iure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.</div>
        </div>
      </li>
      <li class=\\"timeline-item\\">
        <div class=\\"timeline-badge\\"></div>
        <div class=\\"timeline-panel\\">
          <div class=\\"timeline-heading\\">
            <h3 class=\\"h4\\"><span class=\\"glyphicon glyphicon-time\\"></span> 28/03/2021 10:00 - <span class=\\"glyphicon glyphicon-user\\"></span> John Doe</h3>
          </div>
          <div class=\\"timeline-body\\"></div>
        </div>
      </li>
    </ul>
  </div>
</div>"
`;

exports[`http error 1`] = `
"<div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fas fa-ellipsis-v\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div>
<div class=\\"viewport\\"><span class=\\"text-danger\\">Not Found</span></div>"
`;

exports[`no history 1`] = `
"<div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fas fa-ellipsis-v\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div>
<div class=\\"viewport\\">
  <table class=\\"table table-striped\\" horizontal=\\"false\\">
    <thead>
      <tr>
        <th>date</th>
        <th>message</th>
        <th>author</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td colspan=\\"3\\">no_entry</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`two modes with bad default 1`] = `
"<div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div>
<div class=\\"viewport\\">
  <table class=\\"table table-striped\\" horizontal=\\"false\\">
    <thead>
      <tr>
        <th>date</th>
        <th>message</th>
        <th>author</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>06/04/2021 16:23</td>
        <td>
          <p>Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.</p>
        </td>
        <td>John Doe</td>
      </tr>
      <tr>
        <td>03/04/2021 16:23</td>
        <td>
          <p>Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.<br>Sit facilis vitae sapiente. Placeat est qui et.</p>
        </td>
        <td>Alice Smith</td>
      </tr>
      <tr>
        <td>01/04/2021 08:23</td>
        <td>
          <p>Corrupti &lt;strong&gt;reiciendis&lt;/strong&gt; laboriosam repellat adipisci quis.</p>

          <p>Iure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.</p>
        </td>
        <td>Bob Martin</td>
      </tr>
      <tr>
        <td>28/03/2021 10:00</td>
        <td></td>
        <td>John Doe</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`with history 1`] = `
"<div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fas fa-ellipsis-v\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div>
<div class=\\"viewport\\">
  <table class=\\"table table-striped\\" horizontal=\\"false\\">
    <thead>
      <tr>
        <th>date</th>
        <th>message</th>
        <th>author</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>06/04/2021 16:23</td>
        <td>
          <p>Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.</p>
        </td>
        <td>John Doe</td>
      </tr>
      <tr>
        <td>03/04/2021 16:23</td>
        <td>
          <p>Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.<br>Sit facilis vitae sapiente. Placeat est qui et.</p>
        </td>
        <td>Alice Smith</td>
      </tr>
      <tr>
        <td>01/04/2021 08:23</td>
        <td>
          <p>Corrupti &lt;strong&gt;reiciendis&lt;/strong&gt; laboriosam repellat adipisci quis.</p>

          <p>Iure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.</p>
        </td>
        <td>Bob Martin</td>
      </tr>
      <tr>
        <td>28/03/2021 10:00</td>
        <td></td>
        <td>John Doe</td>
      </tr>
    </tbody>
  </table>
</div>"
`;
